reconciler/reconcile_common.go

```diff
@@ -29,12 +30,20 @@ const failedGenerationBump = "NewObservedGenFailure"
 func PreProcessReconcile(ctx context.Context, resource duckv1.KRShaped) {
 	newStatus := resource.GetStatus()
 
-	if newStatus.ObservedGeneration != resource.GetGeneration() {
-		condSet := resource.GetConditionSet()
-		manager := condSet.Manage(newStatus)
-		// Ensure conditions are initialized before we modify.
-		manager.InitializeConditions()
+	// We may be reading a version of the object that was stored at an older version
+	// and may not have had all of the assumed defaults specified.  This won't result
+	// in this getting written back to the API Server, but lets downstream logic make
+	// assumptions about defaulting.
+	if d, ok := resource.(apis.Defaultable); ok {
+		d.SetDefaults(ctx)
+	}
```

Comment:
Even with the comment I fail to understand that part. Why does a function that alters the status need to care about applying defaults? Is the intention to implement some sort of webhook-less defaulting mechanism?

Reply:
No. The webhook is still responsible for defaulting. This is for when defaults change but the user never writes to the resource, the global resync would still default them during reconcile.

You'll see this line repeated in the serving reconcilers:
https://github.com/knative/serving/blob/fcfbb56753d17575969a6f58c432090c8c69733c/pkg/reconciler/configuration/configuration.go#L61

https://github.com/knative/serving/blob/fcfbb56753d17575969a6f58c432090c8c69733c/pkg/reconciler/revision/revision.go#L136

https://github.com/knative/serving/blob/fcfbb56753d17575969a6f58c432090c8c69733c/pkg/reconciler/route/route.go#L112

https://github.com/knative/serving/blob/fcfbb56753d17575969a6f58c432090c8c69733c/pkg/reconciler/service/service.go#L70

I just want to either agree we need this and unify it here if its something we want all reconcilers to be doing. In the future if we can default-on-read instead it may not be nessecary.

Reply:
Got it thanks 👍  I read "revision" instead of "version", that's where my confusion came from.